File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Models/EventDates.swift
import Foundation

/// Dates as the logbook models them: an ISO `yyyy-mm-dd` string, which is what
/// every date column stores and what the API sends.
///
/// A port of the date helpers in `public/app.js` — `todayISO`, `isUpcoming`,
/// `daysUntil`, `fmtCountdown`, `eventDayISO`, `DEFAULT_CHECKLIST` — plus `fmtDate`
/// from `public/js/format.js` (the rest of that file is `LapTime`). Same names, and
/// the same two different clocks the JS uses, deliberately:
///
/// - `todayISO`/`eventDayISO` are **UTC**, because `toISOString()` is.
/// - `daysUntil` counts **local calendar days**, because `new Date(y, m - 1, d)`
///   and `setHours(0,0,0,0)` are local. A countdown has to say "Tomorrow" by the
///   user's midnight, not UTC's.
///
/// Keeping the split means "In 1 day" and "upcoming" agree with the web app on the
/// same data, including in the hours either side of midnight where they differ.
public enum EventDates {
    /// Today's date in UTC, as `yyyy-mm-dd`.
    public static func todayISO(now: Date = Date()) -> String {
        utcISOString(from: now)
    }

    /// A date the user picked, as `yyyy-mm-dd` — in **their** calendar, not UTC.
    ///
    /// This is not `todayISO`'s clock, on purpose, and the difference is a whole day
    /// wide: at 10pm in New York, UTC has already rolled over. A date picker's answer
    /// must round-trip to the day the picker showed, so it is formatted locally —
    /// exactly what the web app's `<input type="date">` submits.
    public static func isoString(from date: Date, calendar: Calendar = .autoupdatingCurrent) -> String {
        let c = calendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    /// The reverse: local midnight on that date, which is what a `DatePicker` binds
    /// to. nil when the string isn't a date.
    public static func date(fromISO iso: String, calendar: Calendar = .autoupdatingCurrent) -> Date? {
        guard let parts = ymd(iso) else { return nil }
        var components = DateComponents()
        components.year = parts.y
        components.month = parts.m
        components.day = parts.d
        return calendar.date(from: components)
    }

    /// Is this event still ahead? String comparison, exactly as the JS: ISO dates
    /// sort lexicographically.
    public static func isUpcoming(_ startDate: String, now: Date = Date()) -> Bool {
        startDate > todayISO(now: now)
    }

    /// Whole local calendar days from today to `iso`. Negative once it's past.
    /// nil when the string isn't a date.
    public static func daysUntil(
        _ iso: String, now: Date = Date(), calendar: Calendar = .autoupdatingCurrent
    ) -> Int? {
        guard let parts = ymd(iso) else { return nil }
        var components = DateComponents()
        components.year = parts.y
        components.month = parts.m
        components.day = parts.d
        guard let target = calendar.date(from: components) else { return nil }
        let today = calendar.startOfDay(for: now)
        return calendar.dateComponents([.day], from: today, to: calendar.startOfDay(for: target)).day
    }

    /// "Today" / "Tomorrow" / "In 5 days". nil when the date won't parse.
    public static func fmtCountdown(
        _ iso: String, now: Date = Date(), calendar: Calendar = .autoupdatingCurrent
    ) -> String? {
        guard let dd = daysUntil(iso, now: now, calendar: calendar) else { return nil }
        if dd <= 0 { return "Today" }
        if dd == 1 { return "Tomorrow" }
        return "In \(dd) days"
    }

    /// ISO date of an event's Nth day — day 1 is `startDate`.
    public static func eventDayISO(_ startDate: String, day: Int) -> String? {
        guard let parts = ymd(startDate) else { return nil }
        var utc = Calendar(identifier: .gregorian)
        utc.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        var components = DateComponents()
        components.year = parts.y
        components.month = parts.m
        components.day = parts.d + day - 1
        guard let date = utc.date(from: components) else { return nil }
        return utcISOString(from: date)
    }

    /// An ISO date rendered for display: "Jul 30, 2026", ordered by the user's
    /// locale. `nil` or an unparseable string is the em dash every other formatter
    /// uses for absent data.
    public static func fmtDate(_ iso: String?) -> String {
        guard let iso, let parts = ymd(iso) else { return "—" }
        var utc = Calendar(identifier: .gregorian)
        utc.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        var components = DateComponents()
        components.year = parts.y
        components.month = parts.m
        components.day = parts.d
        guard let date = utc.date(from: components) else { return "—" }
        // Formatted in UTC too, or a date built at UTC midnight renders as the
        // previous day for anyone west of Greenwich.
        return date.formatted(
            Date.FormatStyle(date: .abbreviated, time: .omitted, timeZone: utc.timeZone)
        )
    }

    /// The prep checklist a new event starts from when the user hasn't made their
    /// own — a port of `DEFAULT_CHECKLIST` in `public/js/checklist.js`, pinned to
    /// it item for item by `contracts/logic/checklist.json`.
    ///
    /// A user who edits the list in Settings gets their own, stored server-side
    /// and served by `GET /api/me` as `User.checklistTemplate`; that wins over
    /// this whenever it is present.
    public static let DEFAULT_CHECKLIST = [
        // First because it is the only item with a deadline days before the event
        // rather than the morning of it.
        "Purchase Track insurance",
        "Tech inspection",
        "Torque lug nuts",
        "Check brake pads & fluid",
        "Set tire pressures",
        "Top off fuel",
        "Pack helmet & gloves",
        "Empty the car — remove loose items",
        "Charge camera / lap timer"
    ]

    // MARK: - Parsing

    /// `yyyy-mm-dd` → components. Rejects anything else, including a date with a
    /// time on it, so a bad string fails visibly rather than becoming 1 January.
    static func ymd(_ iso: String) -> (y: Int, m: Int, d: Int)? {
        let parts = iso.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3,
              parts[0].count == 4,
              let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]),
              (1...12).contains(m), (1...31).contains(d)
        else { return nil }
        return (y, m, d)
    }

    private static func utcISOString(from date: Date) -> String {
        var utc = Calendar(identifier: .gregorian)
        utc.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        return isoString(from: date, calendar: utc)
    }
}

File: apps/android/core/src/main/kotlin/app/trackevolution/core/EventDates.kt
package app.trackevolution.core

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.time.temporal.ChronoUnit

/**
 * Event dates, countdowns, and the prep-checklist default — the Kotlin side of
 * what iOS calls `EventDates`.
 *
 * `LapTime` deliberately skipped the date half of `public/js/format.js` on the
 * grounds that it had no meaning outside a browser. The logbook screens (NS-26)
 * are where it acquires one.
 *
 * **Dates here are calendar dates, never instants.** An event's `start_date` is
 * a date the driver wrote down, not a moment: it is `2026-05-01` at the track
 * regardless of the phone's time zone, and turning it into a timestamp is how
 * "Today" becomes "Tomorrow" for anyone east of UTC. Everything below is
 * [LocalDate] arithmetic for that reason, which also makes it DST-proof for
 * free — a day either side of a transition is still one day.
 */
public object EventDates {

    /** The phone's local calendar date, as the API spells dates. */
    public fun todayIso(today: LocalDate = LocalDate.now()): String = today.toString()

    /**
     * Is this event still ahead? Strictly after today, matching the web's
     * `start_date > todayISO()` — an event that starts today is happening, not
     * upcoming, and the dashboard says so with "Today".
     */
    public fun isUpcoming(startDate: String, today: LocalDate = LocalDate.now()): Boolean =
        parse(startDate)?.isAfter(today) ?: false

    /** Whole days from today to [startDate]; negative once it is past. */
    public fun daysUntil(startDate: String, today: LocalDate = LocalDate.now()): Int? {
        val date = parse(startDate) ?: return null
        return ChronoUnit.DAYS.between(today, date).toInt()
    }

    /** "Today" · "Tomorrow" · "In 5 days", matching the web's wording. */
    public fun fmtCountdown(startDate: String, today: LocalDate = LocalDate.now()): String {
        val days = daysUntil(startDate, today) ?: return EM_DASH
        return when {
            days <= 0 -> "Today"
            days == 1 -> "Tomorrow"
            else -> "In $days days"
        }
    }

    /**
     * "May 1, 2026" — localized, so it follows the phone's region rather than
     * the developer's.
     */
    public fun fmtDate(startDate: String?): String {
        val date = startDate?.let { parse(it) } ?: return EM_DASH
        return date.format(MEDIUM)
    }

    /**
     * The date as a chart x coordinate: days since the epoch.
     *
     * **Days, not an instant.** The web chart plots `new Date(start_date).getTime()`
     * and iOS plots a `timeIntervalSince1970`; both are the same axis scaled
     * differently, and `ChartScale` only ever normalises a range, so the unit is
     * free. Days is the one that cannot drift: a calendar date turned into a
     * timestamp acquires a time zone, and an event would move a day either side of
     * midnight depending on where the phone is. What matters — that a two-year gap
     * between events *looks* like a gap rather than the next point along — is
     * preserved exactly.
     */
    public fun epochDay(startDate: String?): Double? =
        startDate?.let { parse(it) }?.toEpochDay()?.toDouble()

    /** "2 days" / "1 day" / "1.5 days", the way the event tiles read. */
    public fun fmtDays(days: Double): String {
        val whole = days.toInt()
        val text = if (days == whole.toDouble()) whole.toString() else days.toString()
        return "$text ${if (days == 1.0) "day" else "days"}"
    }

    /**
     * The prep list a new event's checklist starts from.
     *
     * Pinned to `public/js/checklist.js` by `contracts/logic/checklist.json`,
     * the same fixture iOS asserts against — two hand-maintained copies of one
     * product decision drift the first time either is edited.
     *
     * Only the fallback: a user who edits the list in Settings gets their own,
     * stored server-side and served by `GET /api/me`.
     */
    public val DEFAULT_CHECKLIST: List<String> = listOf(
        // First because it is the only item with a deadline days before the
        // event rather than the morning of it.
        "Purchase Track insurance",
        "Tech inspection",
        "Torque lug nuts",
        "Check brake pads & fluid",
        "Set tire pressures",
        "Top off fuel",
        "Pack helmet & gloves",
        "Empty the car — remove loose items",
        "Charge camera / lap timer",
    )

    private val MEDIUM: DateTimeFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)

    private const val EM_DASH = "—"

    /** Null rather than throwing: a malformed date must not take out a list. */
    private fun parse(iso: String): LocalDate? = runCatching { LocalDate.parse(iso) }.getOrNull()
}
